core/src/main/java/org/jdbi/v3/core/mapper/reflect/ImmutableMapper.java

```diff
@@ -194,41 +166,44 @@ public T map(final ResultSet rs, final StatementContext ctx) throws SQLException
                 "didn't find any matching columns in result set", type));
         }
 
+        Constructor<MethodHandles.Lookup> constructorTemp = null;
+        try {
+            constructorTemp = MethodHandles.Lookup.class.getDeclaredConstructor(Class.class, int.class);
+            constructorTemp.setAccessible(true);
+        } catch (NoSuchMethodException e) {
+            e.printStackTrace();
+        }
+        final Constructor<MethodHandles.Lookup> constructor = constructorTemp;
+
         return (r, c) -> {
-            Object builder = null;
-            try {
-                builder = builderCreationMethod.invoke(null);
-            } catch (IllegalAccessException | InvocationTargetException e) {
-                throw new MappingException("Could not create Builder", e);
-            }
+            final Map<Method, Object> invocations = new HashMap<>();
             for (int i = 0; i < mappers.size(); i++) {
                 RowMapper<?> mapper = mappers.get(i);
                 PropertyDescriptor property = properties.get(i);
-
                 Object value = mapper.map(r, ctx);
-
-                try {
-                    MethodHandles.lookup().unreflect(property.getWriteMethod()).bindTo(builder).invokeWithArguments(value);
-                } catch (IllegalAccessException e) {
-                    throw new MappingException("Setter for the builder should be public", e);
-                } catch (RuntimeException e) {
-                    throw e;
-                } catch (Throwable e) {
-                    throw new MappingException("Immutable build failed with exception", e);
-                }
+                invocations.put(property.getReadMethod(), value);
             }
-            try {
-                return (T) MethodHandles.lookup().findVirtual(builder.getClass(),
-                    "build",
-                    MethodType.methodType(implementation)).bindTo(builder).invokeWithArguments();
-            } catch (NoSuchMethodException | IllegalAccessException e) {
-                throw new MappingException("Expected to find a Builder with public build() method", e);
-            } catch (Throwable e) {
-                if (e instanceof RuntimeException) {
-                    throw (RuntimeException) e;
+
+            return (T) Proxy.newProxyInstance(this.getClass().getClassLoader(), new Class[]{type}, new InvocationHandler() {
+                private final Map<Method, Object> invocationMap = invocations;
+
+                @Override
+                public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+                    Object returnObj = invocationMap.get(method);
+                    if (returnObj == null) {
+                        if (method.isDefault() && constructor != null) {
```

Comment:
Should be checked for default Values on the evaluation or on the mapping?

Reply:
Definitely on mapping--if you try to wait until the method is invoked, the JDBC result set will be closed.